python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py

```diff
@@ -46,43 +47,60 @@ class FunctionCache(object):
     """
 
     def __init__(self):
-        self._dycode_to_static_func = dict()
-        self._static_func_to_transformer = dict()
+        # Caches the converted static functions. {dygraph_func: static_func}
+        self._converted_static_func_caches = dict()
+        # Caches the converted ast node for same source code.
+        self._code_to_ast_caches = dict()
+        self._dygraph_to_static = DygraphToStaticAst()
 
     def get_or_cache_func(self, func):
-        # code = self._get_dedent_code_string(func)
-        static_func = self._dycode_to_static_func.get(func, None)
+        """
+        Returns the cached static function or converts it when first encounter the function.
+        """
+        static_func = self._converted_static_func_caches.get(func, None)
 
         if static_func is None:
-            static_func, dygraph_to_static_transformer = convert_to_static(func)
-            self._dycode_to_static_func[func] = static_func
-            self._static_func_to_transformer[
-                func] = dygraph_to_static_transformer
+            static_func = self.convert_to_static_with_cache(func)
+            self._converted_static_func_caches[func] = static_func
 
         return static_func
 
-    def get_transformer(self, func):
-        return self._static_func_to_transformer.get(func, None)
+    def convert_to_static_with_cache(self, func):
+        """
+        Converts dygraph function into static function.
+        """
+        # Note: In Python2, it will raise OSError when inspect function
+        # with decorator directly and function.__wrapped__ holds the actual function.
+        func = getattr(func, '__wrapped__', func)
+        source_code = func_to_source_code(func)
+        if source_code in self._code_to_ast_caches:
+            root_wrapper = self._code_to_ast_caches[source_code]
+        else:
+            root = gast.parse(source_code)
+            root_wrapper = self._dygraph_to_static.get_static_ast(root)
+            self._code_to_ast_caches[source_code] = root_wrapper
 
-    def _get_dedent_code_string(self, func):
-        raw_code = inspect.getsource(func)
-        dedent_code = textwrap.dedent(raw_code)
-        return dedent_code
+        # Get static_func from AST
+        static_func, file_name = ast_to_func(root_wrapper.node, func)
+        return static_func
```

Comment:
how about combining these two functions (`get_or_cache_func ` and `convert_to_static_with_cache ` ) into only one function to make logic clearer? And add comment as the follows:
> For the function with same detent source code, only once convert into ast nodes once and cache them.
For two function with same dedent code, the second function will reuse the transformed ast node and save a new tmp file and reload a independent module because they can hold difference global variables in their located files.

Reply:
`get_or_cache_func ` is class method of `FunctionCache` which means it should instance the class first. But `convert_to_static_with_cache ` is provided to shield the implement details of `FunctionCache` that you can call it directly.